src/hotspot/share/code/nmethod.cpp

```diff
@@ -1448,63 +1447,54 @@ nmethod* nmethod::clone(CodeBlobType code_blob_type) {
+  }
+
+  // To make dependency checking during class loading fast, record
+  // the nmethod dependencies in the classes it is dependent on.
+  // This allows the dependency checking code to simply walk the
+  // class hierarchy above the loaded class, checking only nmethods
+  // which are dependent on those classes.  The slow way is to
+  // check every nmethod for dependencies which makes it linear in
+  // the number of methods compiled.  For applications with a lot
+  // classes the slow way is too slow.
+  for (Dependencies::DepStream deps(nm_copy); deps.next(); ) {
+    if (deps.type() == Dependencies::call_site_target_value) {
+      // CallSite dependencies are managed on per-CallSite instance basis.
+      oop call_site = deps.argument_oop(0);
+      MethodHandles::add_dependent_nmethod(call_site, nm_copy);
+    } else {
+      InstanceKlass* ik = deps.context_type();
+      if (ik == nullptr) {
+        continue;  // ignore things like evol_method
+      }
+      // record this nmethod as dependent on this klass
+      ik->add_dependent_nmethod(nm_copy);
+    }
+  }
+
   nm_copy->post_init();
 
+  make_not_used();
+
   return nm_copy;
 }
 
 nmethod* nmethod::relocate_to(nmethod* nm, CodeBlobType code_blob_type) {
-  if (nm == nullptr || !nm->is_relocatable()) {
-    return nullptr;
-  }
-
-  nmethod* nm_copy = nullptr;
-
-  {
-    // Clear inline caches before acquiring any locks
-    VM_ClearNMethodICs clear_nmethod_ics(nm);
-    VMThread::execute(&clear_nmethod_ics);
-
-    MutexLocker mu(CodeCache_lock, Mutex::_no_safepoint_check_flag);
-    nm_copy = nm->clone(code_blob_type);
-
-    if (nm_copy != nullptr) {
-      // To make dependency checking during class loading fast, record
-      // the nmethod dependencies in the classes it is dependent on.
-      // This allows the dependency checking code to simply walk the
-      // class hierarchy above the loaded class, checking only nmethods
-      // which are dependent on those classes.  The slow way is to
-      // check every nmethod for dependencies which makes it linear in
-      // the number of methods compiled.  For applications with a lot
-      // classes the slow way is too slow.
-      for (Dependencies::DepStream deps(nm_copy); deps.next(); ) {
-        if (deps.type() == Dependencies::call_site_target_value) {
-          // CallSite dependencies are managed on per-CallSite instance basis.
-          oop call_site = deps.argument_oop(0);
-          MethodHandles::add_dependent_nmethod(call_site, nm_copy);
-        } else {
-          InstanceKlass* ik = deps.context_type();
-          if (ik == nullptr) {
-            continue;  // ignore things like evol_method
-          }
-          // record this nmethod as dependent on this klass
-          ik->add_dependent_nmethod(nm_copy);
-        }
-      }
+  // Relocate nmethod at safepoint
+  VM_RelocateNMethod relocate_nmethod(nm, code_blob_type);
+  VMThread::execute(&relocate_nmethod);
```

Comment:
You should not do this here. It will be a responsibility of a caller to ensure relocation is done at a safepoint.  A caller will get to a safepoint and relocate a bunch of nmethods.

Reply:
In your PR this will be responsibility of WhiteBox.